File: kotlin-multiplatform/src/commonMain/kotlin/com/ricoh360/thetaclient/capture/Capture.kt
package com.ricoh360.thetaclient.capture

import com.ricoh360.thetaclient.ThetaRepository
import com.ricoh360.thetaclient.transferred.Options
import com.ricoh360.thetaclient.transferred.UnknownResponse
import io.ktor.client.call.body
import io.ktor.client.statement.HttpResponse

/*
 * Capture
 *
 * @property options option of capture
 */
abstract class Capture internal constructor(internal val options: Options) {

    /**
     * Get aperture value.
     *
     * @return aperture value
     */
    fun getAperture() = options.aperture?.let {
        ThetaRepository.ApertureEnum.get(it)
    }

    /**
     * Get color temperature of the camera (Kelvin).
     *
     * @return Color temperature
     */
    fun getColorTemperature() = options._colorTemperature

    /**
     * Get exposure compensation (EV).
     *
     * @return Exposure compensation (EV)
     */
    fun getExposureCompensation() = options.exposureCompensation?.let {
        ThetaRepository.ExposureCompensationEnum.get(it)
    }

    /**
     * Get operating time (sec.) of the self-timer.
     *
     * @return Operating time (sec.) of the self-timer
     */
    fun getExposureDelay() = options.exposureDelay?.let {
        ThetaRepository.ExposureDelayEnum.get(it)
    }

    /**
     * Get exposure program.
     *
     * @return Exposure program
     */
    fun getExposureProgram() = options.exposureProgram?.let {
        ThetaRepository.ExposureProgramEnum.get(it)
    }

    /**
     * Get GPS information.
     *
     * @return GPS information
     */
    fun getGpsInfo() = options.gpsInfo?.let {
        ThetaRepository.GpsInfo(it)
    }

    /**
     * Get turns position information assigning ON/OFF.
     *
     * @return Turns position information assigning ON/OFF
     */
    fun getGpsTagRecording() = options._gpsTagRecording?.let {
        ThetaRepository.GpsTagRecordingEnum.get(it)
    }

    /**
     * Set ISO sensitivity.
     *
     * It can be set for video shooting mode at RICOH THETA V firmware v3.00.1 or later.
     * Shooting settings are retained separately for both the Still image shooting mode and Video shooting mode.
     *
     * When the exposure program (exposureProgram) is set to Manual or ISO Priority
     *
     * @return ISO sensitivity
     */
    fun getIso() = options.iso?.let {
        ThetaRepository.IsoEnum.get(it)
    }

    /**
     * Get ISO sensitivity upper limit when ISO sensitivity is set to automatic.
     *
     * @return ISO sensitivity upper limit
     */
    fun getIsoAutoHighLimit() = options.isoAutoHighLimit?.let {
        ThetaRepository.IsoAutoHighLimitEnum.get(it)
    }

    /**
     * Get white balance.
     *
     * @return White balance
     */
    fun getWhiteBalance() = options.whiteBalance?.let {
        ThetaRepository.WhiteBalanceEnum.get(it)
    }

    /*
     * Builder
     */
    abstract class Builder<T> {
        internal val options = Options()

        /**
         * Set aperture value.
         *
         * @param aperture aperture value
         * @return Builder
         */
        fun setAperture(aperture: ThetaRepository.ApertureEnum): T {
            options.aperture = aperture.value
            @Suppress("UNCHECKED_CAST")
            return this as T
        }

        /**
         * Set color temperature of the camera (Kelvin).
         *
         * 2500 to 10000. In 100-Kelvin units.
         *
         * @param kelvin Color temperature
         * @return Builder
         */
        fun setColorTemperature(kelvin: Int): T {
            options._colorTemperature = kelvin
            @Suppress("UNCHECKED_CAST")
            return this as T
        }

        /**
         * Set exposure compensation (EV).
         *
         * @param value Exposure compensation
         * @return Builder
         */
        fun setExposureCompensation(value: ThetaRepository.ExposureCompensationEnum): T {
            options.exposureCompensation = value.value
            @Suppress("UNCHECKED_CAST")
            return this as T
        }

        /**
         * Set operating time (sec.) of the self-timer.
         *
         * @param delay Operating time (sec.)
         * @return Builder
         */
        fun setExposureDelay(delay: ThetaRepository.ExposureDelayEnum): T {
            options.exposureDelay = delay.sec
            @Suppress("UNCHECKED_CAST")
            return this as T
        }

        /**
         * Set exposure program. The exposure settings that take priority can be selected.
         *
         * @param program Exposure program
         * @return Builder
         */
        fun setExposureProgram(program: ThetaRepository.ExposureProgramEnum): T {
            options.exposureProgram = program.value
            @Suppress("UNCHECKED_CAST")
            return this as T
        }

        /**
         * Set GPS information.
         *
         * @param gpsInfo GPS information
         * @return Builder
         */
        fun setGpsInfo(gpsInfo: ThetaRepository.GpsInfo): T {
            options.gpsInfo = gpsInfo.toTransferredGpsInfo()
            @Suppress("UNCHECKED_CAST")
            return this as T
        }

        /**
         * Set turns position information assigning ON/OFF.
         *
         * For RICOH THETA X
         *
         * @param value Turns position information assigning ON/OFF
         * @return Builder
         */
        fun setGpsTagRecording(value: ThetaRepository.GpsTagRecordingEnum): T {
            options._gpsTagRecording = value.value
            @Suppress("UNCHECKED_CAST")
            return this as T
        }

        /**
         * Set ISO sensitivity.
         *
         * It can be set for video shooting mode at RICOH THETA V firmware v3.00.1 or later.
         * Shooting settings are retained separately for both the Still image shooting mode and Video shooting mode.
         *
         * When the exposure program (exposureProgram) is set to Manual or ISO Priority
         *
         * @param iso ISO sensitivity
         * @return Builder
         */
        fun setIso(iso: ThetaRepository.IsoEnum): T {
            options.iso = iso.value
            @Suppress("UNCHECKED_CAST")
            return this as T
        }

        /**
         * Set ISO sensitivity upper limit when ISO sensitivity is set to automatic.
         *
         * @param iso ISO sensitivity upper limit
         * @return Builder
         */
        fun setIsoAutoHighLimit(iso: ThetaRepository.IsoAutoHighLimitEnum): T {
            options.isoAutoHighLimit = iso.value
            @Suppress("UNCHECKED_CAST")
            return this as T
        }

        /**
         * Set white balance.
         *
         * It can be set for video shooting mode at RICOH THETA V firmware v3.00.1 or later.
         * Shooting settings are retained separately for both the Still image shooting mode and Video shooting mode.
         *
         * @param whiteBalance White balance
         * @return Builder
         */
        fun setWhiteBalance(whiteBalance: ThetaRepository.WhiteBalanceEnum): T {
            options.whiteBalance = whiteBalance.value
            @Suppress("UNCHECKED_CAST")
            return this as T
        }
    }
}

internal suspend fun isCanceledShootingResponse(httpResponse: HttpResponse): Boolean {
    try {
        val response: UnknownResponse = httpResponse.body()
        if (response.error?.isCanceledShootingCode() == true) {
            return true
        }
    } catch (_: Exception) {
    }
    return false
}

File: react-native/src/capture/capture.ts
import type {
  ApertureEnum,
  ExposureCompensationEnum,
  ExposureDelayEnum,
  ExposureProgramEnum,
  GpsInfo,
  GpsTagRecordingEnum,
  IsoAutoHighLimitEnum,
  IsoEnum,
  Options,
  PhotoFileFormatEnum,
  WhiteBalanceEnum,
} from '../theta-repository/options';

/**
 * Capture Builder class
 */
export abstract class CaptureBuilder<T extends CaptureBuilder<T>> {
  /** capture options */
  options: Options;

  /** construct capture builder instance */
  constructor() {
    this.options = {};
  }

  /**
   * Set aperture value.
   * @param {ApertureEnum} aperture aperture value to set
   * @return CaptureBuilder
   */
  setAperture(aperture: ApertureEnum): T {
    this.options.aperture = aperture;
    return this as unknown as T;
  }

  /**
   * Set color temperature of the camera (Kelvin).
   * @param {number} kelvin Color temperature to set
   * @return CaptureBuilder
   */
  setColorTemperature(kelvin: number): T {
    this.options.colorTemperature = kelvin;
    return this as unknown as T;
  }

  /**
   * Set exposure compensation (EV).
   * @param {ExposureCompensationEnum} value Exposure compensation to set
   * @return CaptureBuilder
   */
  setExposureCompensation(value: ExposureCompensationEnum): T {
    this.options.exposureCompensation = value;
    return this as unknown as T;
  }

  /**
   * Set operating time (sec.) of the self-timer.
   * @param {ExposureDelayEnum} delay Operating time to set
   * @return CaptureBuilder
   */
  setExposureDelay(delay: ExposureDelayEnum): T {
    this.options.exposureDelay = delay;
    return this as unknown as T;
  }

  /**
   * Set exposure program. The exposure settings that take priority can be selected.
   * @param {ExposureProgramEnum} program Exposure program to set
   * @return CaptureBuilder
   */
  setExposureProgram(program: ExposureProgramEnum): T {
    this.options.exposureProgram = program;
    return this as unknown as T;
  }
  /**
   * Set GPS information.
   * @param {GpsInfo} gpsInfo GPS information to set
   * @return CaptureBuilder
   */
  setGpsInfo(gpsInfo: GpsInfo): T {
    this.options.gpsInfo = gpsInfo;
    return this as unknown as T;
  }

  /**
   * Set turns position information assigning ON/OFF.
   * @param {GpsTagRecordingEnum} value Turns position information assigning ON/OFF to set
   * @return CaptureBuilder
   */
  setGpsTagRecording(value: GpsTagRecordingEnum): T {
    this.options._gpsTagRecording = value;
    return this as unknown as T;
  }

  /**
   * Set ISO sensitivity.
   * @param {IsoEnum} iso ISO sensitivity to set
   * @return CaptureBuilder
   */
  setIso(iso: IsoEnum): T {
    this.options.iso = iso;
    return this as unknown as T;
  }

  /**
   * Set ISO sensitivity upper limit when ISO sensitivity is set to automatic.
   * @param {IsoAutoHighLimitEnum} iso ISO sensitivity upper limit to set
   * @return CaptureBuilder
   */
  setIsoAutoHighLimit(iso: IsoAutoHighLimitEnum): T {
    this.options.isoAutoHighLimit = iso;
    return this as unknown as T;
  }

  /**
   * Set white balance.
   * @param {WhiteBalanceEnum} whiteBalance White balance to set
   * @return CaptureBuilder
   */
  setWhiteBalance(whiteBalance: WhiteBalanceEnum): T {
    this.options.whiteBalance = whiteBalance;
    return this as unknown as T;
  }
}

/** Capturing status */
export const CapturingStatusEnum = {
  /** The process is starting */
  STARTING: 'STARTING',
  /** Capture in progress */
  CAPTURING: 'CAPTURING',
  /** Self-timer in progress */
  SELF_TIMER_COUNTDOWN: 'SELF_TIMER_COUNTDOWN',
} as const;

/** type definition of CapturingStatusEnum */
export type CapturingStatusEnum =
  (typeof CapturingStatusEnum)[keyof typeof CapturingStatusEnum];

export class PhotoCaptureBuilderBase<
  T extends PhotoCaptureBuilderBase<T>
> extends CaptureBuilder<T> {
  /**
   * Set photo file format.
   * @param {PhotoFileFormatEnum} fileFormat Photo file format to set
   * @return PhotoCaptureBuilder
   */
  setFileFormat(fileFormat: PhotoFileFormatEnum): T {
    this.options.fileFormat = fileFormat;
    return this as unknown as T;
  }
}
